### ISSUE TITLE
add capability to have different solution intervals per direction

### ISSUE THREAD

**Issue by o-smirnov** [ratt-ru/CubiCal#84]
We may need to solve towards some directions with a different time/freq interval than towards others. @JSKenyon, I think I see an elegant way to implement this with a minor tweak of the current kernels:

* We keep a "fundamantal" solution interval (what is now referred to as ``f_int`` and ``t_int`` throughout the code): these are the lowest common denominators (LCDs) of all the directions' intervals. In the extreme case, this is just 1x1. Each direction's actual solution interval is then some integer multiple of this, say N\*f_int and M\*t_int. The N,M factors are different per direction.

* Basically, we want to have the same update rule applied to every NxM block of "fundamental" intervals, right? So in the kernel where you compute J^H.r and J^H.J, you simply take an additional sum along the time and frequency axes (see e.g. https://github.com/ratt-ru/CubiCal/blob/master/cubical/kernels/cyfull_complex.pyx#L161): that is, you sum every block of NxM entries together, and assign the sum to every element in the block.

Does this make sense? Is it as easy as it seems to me?

**Comment by JSKenyon** [comment]
It is possible that it may be that simple @o-smirnov. I have not given it much thought. I would be concerned about the possible impact on performance - it might be negligible, but would require testing. It would also have some impact on the accumulation operations in the kernel. Currently, we simply accumulate into the appropriate element in the (n_tint, n_fint) dimensions. It is true that we could assign the sum to every element in a (dir_t_int, dir_f_int) block, but that would need to occur after the relevant accumulation. Perhaps I have misunderstood though - given how easy it seems to you.

**Comment by o-smirnov** [comment]
*It is true that we could assign the sum to every element in a (dir_t_int, dir_f_int) block, but that would need to occur after the relevant accumulation.*

Exactly. Easy, right? Bet you a beer the performance hit will be negligible. Especially compared to the opportunities this opens up for improving DD-cal!

**Comment by JSKenyon** [comment]
I think the only other concern will be the book keeping - perhaps the best approach in this case is to just assume an LCD of 1 if you have variable solution intervals, and not worry about an intermediate (t_int_lcd, f_int_lcd) shaped array. Will think about that. Then you just work with full tf-resolution gains which the gains repeated over the relevant intervals. Thoughts @o-smirnov ?

**Comment by o-smirnov** [comment]
That'll do for a start, but let's think about it a bit more. There of course you are giving up some serious performance -- especially if the GCD (not LCD, my mistake) is quite large.

**Comment by JSKenyon** [comment]
Yeah, there there may be a bit of a performance hit - definitely a memory footprint bump.

**Comment by IanHeywood** [comment]
```
Each direction's actual solution interval is then some integer multiple of this, say N*f_int and M*t_int.
```

Do N and M have to be integers?

**Comment by o-smirnov** [comment]
Yes. But f_int and t_int can be 1.:)

Cheers,
Oleg

---
Sent from my phone. Quality of spelling inversely proportional to finger
size.

**Comment by IanHeywood** [comment]
"Hmmmm."

**Comment by JSKenyon** [comment]
![img_1399rot](https://user-images.githubusercontent.com/6582745/32053245-09c03b4c-ba5b-11e7-971d-b989e60d5104.JPG)

Putting this very old school, handwritten rambling here. Just trying to make sure I have all the threads nailed down before I try implement something.

**Comment by o-smirnov** [comment]
Yeah looks right to me.

**Comment by JSKenyon** [comment]
I have been playing around, trying to find the simplest way of mapping from a partially complete array to one which is correct at the GCD resolution. Here is what I have so far. 

```
import numpy as np
from fractions import gcd
from math import ceil

n_dir, n_tim, n_fre, n_ant = 2, 10, 10, 3

t_ints = [4,2]
f_ints = [3,6]

assert len(t_ints) == len(f_ints) == n_dir

# Find the GCD for time and frequency.

t_gcd = reduce(gcd, t_ints)
f_gcd = reduce(gcd, f_ints)

# This reduces an array to the GCD dimensions. Shouldnt be necessary in CubiCal.

# data = np.random.random([n_dir, n_tim, n_fre, n_ant, n_ant, 2, 2])

# t_red = range(0, n_tim, t_gcd)
# f_red = range(0, n_fre, f_gcd)

# gcd_data = np.add.reduceat(np.add.reduceat(data, t_red, axis=1), f_red, axis=2)

# Figure out the number of intervals w.r.t. the GCD.

n_gcd_t = ceil(float(n_tim)/t_gcd)
n_gcd_f = ceil(float(n_fre)/f_gcd)

# Create a gain array filled with random values - meaningless, purely for demonstration.

gains = np.random.random([n_dir, n_gcd_t, n_gcd_f, n_ant, 2, 2])

# Set up a mapping from a partially incomplete array into an array where elements are 
# repeated to be consistent with intervals. Once off per GM/chunk in CubiCal.

t_mapping = np.zeros([n_dir, n_gcd_t, n_gcd_f], dtype=np.int)
f_mapping = np.zeros([n_dir, n_gcd_t, n_gcd_f], dtype=np.int)

for d in xrange(n_dir):

    t_ind = np.arange(n_gcd_t)*t_gcd//t_ints[d] 
    f_ind = np.arange(n_gcd_f)*f_gcd//f_ints[d]

    t_mapping[d, ...], f_mapping[d, ...] = np.meshgrid(t_ind, f_ind, indexing="ij")
 
# Applies the mapping. Loop over direction seems to be a necessary evil as our mapping
# is different per direction.

for d in xrange(n_dir):

    gains[d, ...] = gains[d, t_mapping[d, ...], f_mapping[d, ...], ...]

```

**Comment by JSKenyon** [comment]
Fiddled a little more, and found an alternative of the mapping:

```
# An alternative approach which doesn't create a meshgrid. Slightly faster.

indices = []

for d in xrange(n_dir):

    t_ind = np.arange(n_gcd_t)*t_gcd//t_ints[d] 
    f_ind = np.arange(n_gcd_f)*f_gcd//f_ints[d]

    [indices.extend([d*n_gcd_t*n_gcd_f + t*n_gcd_f + f]) for t in t_ind for f in f_ind]

# Applies the mapping. No loop required - accomplished using reshape.

gains = gains.reshape(-1, n_ant, 2, 2)[indices, ...].reshape(n_dir, n_gcd_t, n_gcd_f, n_ant, 2, 2)
```

The indices list could be cashed on the GM so that this operation be performed with relative ease.

**Comment by o-smirnov** [comment]
Yeah looks good -- the reshape() approach is actually what I converged onto when I was doing this in MeqTrees stefcal...
